lib/src/validator/pubspec_typo.dart

```diff
@@ -0,0 +1,73 @@
+// Copyright (c) 2020, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import '../entrypoint.dart';
+import '../levenshtein.dart';
+import '../validator.dart';
+
+/// Validates that a package's pubspec does not contain any typos in its keys.
+class PubspecTypoValidator extends Validator {
+  PubspecTypoValidator(Entrypoint entrypoint) : super(entrypoint);
+
+  @override
+  Future validate() async {
+    final fields = entrypoint.root.pubspec.fields;
+
+    if (fields == null) return;
+
+    for (final key in fields.keys) {
+      var bestLevenshteinRatio = 100.0;
+      var closestKey = '';
+
+      for (final validKey in _validPubspecKeys) {
+        /// Use a ratio to account allow more more typos in strings with
+        /// longer lengths.
+        final ratio =
+            levenshteinDistance(key, validKey) / (validKey.length + key.length);
+        if (ratio < bestLevenshteinRatio) {
+          bestLevenshteinRatio = ratio;
+          closestKey = validKey;
+        }
+      }
+
+      /// 0.21 is a magic value determined by looking at the most common typos
+      /// in all the pubspecs on pub.dev.
+      if (bestLevenshteinRatio > 0 && bestLevenshteinRatio < 0.21) {
+        warnings.add('"$key" is not a key recognized by pub - '
+            'did you mean "$closestKey"?');
+      }
```

Comment:
Maybe I suggest that we add at-most 3 typo warnings? So as to not drown all the other warnings?

We could even go as far as to say, we only have a warning for the closest typo, and not for more than one.
WDYT?

I'm mostly concerned we could be drowning all other warnings.

Reply:
hmm because this picks up typos, and we [experimentally validated](https://gist.github.com/walnutdust/5d0f151025fc986c5606d87fa15d5eb5) with real pubspecs that all of the other keys are fine (i.e. no false positives),  it doesn't feel like there will be too many typo warnings.

I'll do what you suggested and add at most 3 just in case!